### PR TITLE
Losing movement on combat or failed evade

### PR DESCRIPTION
**Pull request by petercombs**
Implement by character having an avoid_monsters attribute that gets reset on Upkeep. This can also be used to implement not fighting monsters upon emergence from a gate.

### CODE REVIEW

File: eldritch/monsters.py [hunk 1]
```diff
@@ -7,7 +9,7 @@ def json_repr(self):
     return self.name
 
 
-class Monster(object):
+class Monster(Asset):
```
Comment (petercombs) [inline eldritch/monsters.py:12]: I don't remember why I did this... I think it had something to do with accessing the monster's name. Regardless, it seems more consistent with the rest of the codebase.

File: eldritch/events.py [hunk 2]
```diff
@@ -1511,6 +1533,8 @@ def resolve(self, state):
       return False
     if self.check.successes >= self.monster.toughness(state):
       # TODO: take the monster as a trophy
+      # Stand-in until we implement trophy code to allow MoveMultipleThroughMonster test to work
+      self.monster.place = None
```
Comment (fingoldfish) [inline eldritch/events.py:1537]: You set up some trophy code above, inside of the Combat event?

File: eldritch/test_events.py [hunk 3]
```diff
@@ -107,6 +107,34 @@ def testMoveOneSpace(self):
...
+
+    third_choice = self.resolve_to_choice(CombatChoice)
+    self.assertIsNone(third_choice.choices)
+    third_choice.resolve(self.state, [])
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=5)):
+      self.resolve_until_done()
+
+    self.assertTrue(movement.is_resolved())
+    self.assertEqual(self.char.place.name, "Easttown")
+    # self.assertEqual(self.char.movement_points, 0)
```
Comment (fingoldfish) [inline eldritch/test_events.py:134]: Why are you not asserting on the movement points here? I know monster trophies may not be complete, but you set movement points to 0 correctly, right?

File: eldritch/test_events.py [hunk 4]
```diff
@@ -120,6 +148,123 @@ def testMoveMultipleSpaces(self):
...
+
+    self.state.event_stack.append(movement)
+    self.resolve_until_done()
+
+    self.assertTrue(movement.is_resolved())
+    self.assertEqual(self.char.place.name, "Easttown")
+    self.assertEqual(self.char.movement_points, 0)
+
+  def testMoveMultipleThroughMonster(self):
+    self.char.fight_will_slider = 2
```
Comment (fingoldfish) [inline eldritch/test_events.py:166]: Why did you change the fight_will_slider if you're going to have them pass the combat check anyway?

File: eldritch/test_events.py [hunk 5]
```diff
@@ -120,6 +148,123 @@ def testMoveMultipleSpaces(self):
...
+    self.assertTrue(movement.is_resolved())
+    self.assertEqual(self.char.place.name, "Easttown")
+    # self.assertEqual(self.char.movement_points, 0)
+    # TODO: ^ should this pass?
+    # self.assertIn(cultist, self.char.possessions)
+    # TODO: take the monster as a trophy
+
+  def testMoveMultipleThroughMonsterFailedEvade(self):
+    self.char.fight_will_slider = 2
+    self.char.speed_sneak_slider = 0
```
Comment (fingoldfish) [inline eldritch/test_events.py:198]: You're setting the speed_sneak_slider to 0, but then asserting that the character has 4 movement points. A bit weird, though I suppose it works if you need the extra sneak to make the test work.

File: eldritch/test_events.py [hunk 6]
```diff
@@ -120,6 +148,123 @@ def testMoveMultipleSpaces(self):
...
+    third_choice = self.resolve_to_choice(CombatChoice)
+    self.assertIsNone(third_choice.choices)
+    third_choice.resolve(self.state, [])
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=5)):
+      self.resolve_until_done()
+
+    self.assertTrue(movement.is_resolved())
+    self.assertEqual(self.char.place.name, "Easttown")
+    # self.assertEqual(self.char.movement_points, 0)
+    # TODO: ^ should this pass?
```
Comment (fingoldfish) [inline eldritch/test_events.py:192]: Okay, I figured this out. When you run the unit tests, it actually starts with turn_phase == "upkeep". That means that when you use EndMovement (a subclass of EndTurn), it actually advances from upkeep to movement. When it advances from upkeep to movement, it resets the character's movement points.

I recommend starting these tests in the movement phase.

**Review by fingoldfish — CHANGES_REQUESTED** [review]

File: eldritch/events.py [hunk 7]
```diff
@@ -121,6 +121,10 @@ class EndUpkeep(EndTurn):
   def __init__(self, character):
     super(EndUpkeep, self).__init__(character, "upkeep")
 
+  def __resolve__(self, state):
```
[the remark this answers is not in the document]
Comment (petercombs) [reply eldritch/events.py:124]: Yep.

File: eldritch/events.py (hunk 2 above)
Comment (petercombs) [reply eldritch/events.py:1537]: Yes, though that code is wrong too. Probably TakeAsTrophy (or similar) should be an event, so we can hijack it for things like Loot Piñatas, personal stories, etc. As a wise man once said, "Passing information between events is most of what happens in this game."

Probably there should only be one bit where I move the monster, though. Any preferences on it appearing in Combat or CombatRound?

File: eldritch/events.py [hunk 8]
```diff
@@ -168,26 +172,33 @@ def __init__(self, character, dest):
...
   def start_str(self):
-    return ""
+    return f"{self.character.name} moving from {self.character.place.name}"
 
   def finish_str(self):
-    return f"{self.character.name} moved to {self.dest.name}"
+    if self.moved:
+      return f"{self.character.name} moved to {self.dest.name}"
+    else:
+      return f"{self.character.name} did not have enough movement points, and stayed in {self.character.place.name}"
```
[the remark this answers is not in the document]
Comment (petercombs) [reply eldritch/events.py:201]: I'm shortening it in case the location is closed, which would also block movement.

File: eldritch/test_events.py (hunk 5 above)
Comment (petercombs) [reply eldritch/test_events.py:198]: Nah, that was mostly me trying to make changes to it, then finding that it did nothing because of your observation below that we weren't actually in the movement phase.

File: eldritch/events.py (hunk 2 above)
Comment (fingoldfish) [reply eldritch/events.py:1536]: I think either Combat or CombatRound works. I can't think of a reason why one would work better than the other.

File: eldritch/monsters.py [hunk 9]
```diff
@@ -28,7 +30,8 @@ def __init__(self, name, movement, dimension, difficulties, damages, toughness,
     assert not damages.keys() - self.DAMAGES
     assert not damages.keys() - difficulties.keys()
     assert not set(attributes) - self.ATTRIBUTES
-    self.name = name
+    # self.name = name
```
Comment (fingoldfish) [inline eldritch/monsters.py:33]: Comment can be removed.

File: eldritch/test_events.py [hunk 10]
```diff
@@ -111,6 +111,36 @@ def testMoveOneSpace(self):
     self.assertEqual(self.char.place.name, "Easttown")
     self.assertEqual(self.char.movement_points, 3)
 
+  def testMoveOneSpaceToMonster(self):
+    while self.state.turn_phase != "movement":
+      self.state.next_turn()
+    maniac = next(monster for monster in self.state.monsters if monster.name == "Maniac")
+    maniac.place = self.state.places["Easttown"]
+    movement = Movement(self.char, [self.state.places["Easttown"]])
```
Comment (fingoldfish) [inline eldritch/test_events.py:119]: You'll have to rewrite this to use CityMovement - sorry.

File: eldritch/events.py [hunk 11]
```diff
@@ -1471,8 +1490,10 @@ def resolve(self, state):
       return False
     if self.check.successes >= 1:
       self.evaded = True
+      self.character.avoid_monsters = True
```
[the remark this answers is not in the document]
Comment (fingoldfish) [reply eldritch/events.py:1492]: Upon further review, I definitely think this belongs in EvadeOrFightAll - there are some monsters where you can lose combat against them instead of successfully evading them.

File: eldritch/test_events.py (hunk 10 above)
Comment (petercombs) [reply eldritch/test_events.py]: I like CityMovement better now that we have advance_to_phase!

**Comment by petercombs** [comment]
Not sure I understood the comment above about "Upon further review, I definitely think this belongs in EvadeOrFightAll - there are some monsters where you can lose combat against them instead of successfully evading them." Does the new method of adding monsters to an ignore_monsters list work better?

File: eldritch/monsters.py [hunk 12]
```diff
@@ -50,6 +53,9 @@ def __init__(
       self._attributes.add("nightmarish")
     self.place = None
 
+  def __repr__(self):
+    return f"<Monster: {self.name} {id(self)} at {self.place}>"
```
Comment (fingoldfish) [inline eldritch/monsters.py:57]: Probably self.place.name if self.place else nowhere

File: eldritch/eldritch.py [hunk 13]
```diff
@@ -358,7 +358,11 @@ def clear_usables(self):
   def get_interrupts(self, event):
     interrupts = []
     if isinstance(event, events.MoveOne):
-      nearby_monsters = [mon for mon in self.monsters if mon.place == event.character.place]
+      nearby_monsters = [
+          mon
+          for mon in self.monsters
+          if mon.place == event.character.place and mon not in event.character.avoid_monsters
```
Comment (fingoldfish) [inline eldritch/eldritch.py:364]: I notice you are doing the filtering for character.avoid_monsters both here before you create the EvadeOrFightAll event, and inside the `__init__` of EvadeOrFightAll. You probably only need one?

File: eldritch/events.py [hunk 14]
```diff
@@ -2423,9 +2442,11 @@ def resolve(self, state):
       return False
     if self.check.successes >= 1:
       self.evaded = True
+      self.character.avoid_monsters.append(self.monster)
       return True
     self.damage = Loss(
         self.character, {"stamina": self.monster.damage("combat", state, self.character)})
+    self.character.movement_points = 0
```
Comment (fingoldfish) [inline eldritch/events.py:2449]: Nitpick: I would either put this line above or below the use of self.damage so that you don't split the definition from the place where it's appended to the stack.

File: eldritch/events.py [hunk 15]
```diff
@@ -2438,7 +2459,8 @@ def start_str(self):
   def finish_str(self):
     if self.evaded:
       return f"{self.character.name} evaded a {self.monster.name}"
-    return f"{self.character.name} did not evade the {self.monster.name}"
+    return (f"{self.character.name} did not evade the {self.monster.name}"
+            + "and lost any remaining movement")
```
Comment (fingoldfish) [inline eldritch/events.py:2463]: Nit: You will want to add a space between the monster name and the word `and`.

File: eldritch/events.py (hunk 2 above)
Comment (fingoldfish) [reply eldritch/events.py:2504]: So I realize I said last time that you could do this in either Combat or CombatRound. And I also realize this is a placeholder. But at a minimum, the code for taking the trophy (adding it the player's possessions) and the code for setting the monster's place to None should live in the same place - otherwise someone is going to be very confused in the future.

File: eldritch/test_events.py [hunk 16]
```diff
@@ -249,6 +250,39 @@ def testMoveOneSpace(self):
...
+    next_choice = self.resolve_to_choice(MultipleChoice)
+    self.assertEqual(next_choice.choices, ["Flee", "Fight"])
+    next_choice.resolve(self.state, "Fight")
+
+    third_choice = self.resolve_to_choice(CombatChoice)
+    self.assertFalse(third_choice.choices)
+    third_choice.resolve(self.state, [])
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=5)):
+      movement = self.resolve_to_choice(CityMovement)
+      self.assertFalse(movement.choices)
```
Comment (fingoldfish) [inline eldritch/test_events.py:277]: Nit: the assertFalse and movement resolution can go outside of the patch block.

File: eldritch/test_events.py [hunk 17]
```diff
@@ -262,6 +296,151 @@ def testMoveMultipleSpaces(self):
...
+    )
+    self.assertFalse(movement.is_resolved())
+    self.assertEqual(self.char.movement_points, 1)
+
+    self.state.event_stack.append(movement)
+    self.resolve_until_done()
+
+    self.assertTrue(movement.is_resolved())
+    self.assertEqual(self.char.place.name, "Easttown")
+    self.assertEqual(self.char.movement_points, 0)
```
Comment (fingoldfish) [inline eldritch/test_events.py:313]: Since you added the `moved` attribute in this PR, maybe also assert that the `moved` attribute gets set correctly?

File: eldritch/test_events.py [hunk 18]
```diff
@@ -262,6 +296,151 @@ def testMoveMultipleSpaces(self):
...
+
+  def testMoveMultipleThroughMonsterFight(self):
```
Comment (fingoldfish) [inline eldritch/test_events.py:315]: Is the key difference between this test and testMoveOneSpaceToMonster that this one starts with the character moving multiple spaces? Maybe add a comment to this test.

File: eldritch/test_events.py [hunk 19]
```diff
@@ -262,6 +296,151 @@ def testMoveMultipleSpaces(self):
...
+    movement.resolve(self.state, "Graveyard")
+
+    choice = self.resolve_to_choice(MultipleChoice)
+
+    self.assertEqual(choice.choices, ["Fight", "Evade"])
+    choice.resolve(self.state, "Evade")
+
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=3)):
+      next_choice = self.resolve_to_choice(MultipleChoice)
+    self.assertEqual(sorted(next_choice.choices), ["Fight", "Flee"])
```
Comment (fingoldfish) [inline eldritch/test_events.py:382]: Pro tip: you can use the handy assertCountEqual and not worry about the order.

File: eldritch/test_events.py [hunk 20]
```diff
@@ -262,6 +296,151 @@ def testMoveMultipleSpaces(self):
...
+    next_choice.resolve(self.state, "Flee")
+
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=5)):
+      movement = self.resolve_to_choice(CityMovement)
+
+    self.assertFalse(movement.choices)
+    movement.resolve(self.state, movement.none_choice)
+    self.resolve_until_done()
```
Comment (fingoldfish) [inline eldritch/test_events.py:390]: You should add a comment here - using resolve_until_done() also tests that you don't have to evade the zombie again when you're done with your movement.

File: eldritch/test_events.py [hunk 21]
```diff
@@ -262,6 +296,151 @@ def testMoveMultipleSpaces(self):
...
+    self.assertIn("Rivertown", movement.choices)
+    movement.resolve(self.state, "Rivertown")
+    choice = self.resolve_to_choice(MultipleChoice)
+    # TODO: Choose which one to evade first
+    # For now, Zombie is the first
+    self.assertEqual(choice.choices, ["Fight", "Evade"])
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=4)):
+      choice.resolve(self.state, "Evade")
+      next_choice = self.resolve_to_choice(MultipleChoice)
+      self.assertEqual(sorted(next_choice.choices), ["Fight", "Flee"])
```
Comment (fingoldfish) [inline eldritch/test_events.py:425]: Same nitpick - can put the asserts outside of the patch block.

File: eldritch/test_events.py [hunk 22]
```diff
@@ -262,6 +296,151 @@ def testMoveMultipleSpaces(self):
...
+
+    self.assertFalse(movement.choices)
+    movement.resolve(self.state, movement.none_choice)
+    self.assertTrue(movement.is_resolved())
+    self.assertEqual(self.char.place.name, "Easttown")
+    self.assertEqual(self.char.stamina, 4)
+    self.assertEqual(self.char.sanity, 5)
+    self.assertEqual(self.char.movement_points, 0)
+    # TODO: ^ should this pass?
+
```
Comment (fingoldfish) [inline eldritch/test_events.py:443]: Let's add a TODO for one more test - if the character has a motorcycle, they should not be able to exhaust the motorcycle to get back any movement points after entering combat. I haven't thought too deeply about how to do that yet - maybe set movement points to -float("inf") instead of 0?

**Review by fingoldfish — CHANGES_REQUESTED** [review]

File: eldritch/test_events.py (hunk 22 above)
Comment (fingoldfish) [reply eldritch/test_events.py:443]: Okay, one more TODO:
a test where you fight a dream flier in a location with a gate, lose, get pulled through the gate, immediately cast Find Gate, return, and then don't have to fight again.

File: eldritch/test_events.py (hunk 19 above)
Comment (petercombs) [reply eldritch/test_events.py:382]: Done

File: eldritch/monsters.py (hunk 12 above)
Comment (fingoldfish) [reply eldritch/monsters.py:59]: Should still be self.place.name instead of self.place

File: eldritch/eldritch.py (hunk 13 above)
Comment (fingoldfish) [reply eldritch/eldritch.py:364]: Same comment still applies. I think EvadeOrFightAll probably works better, since you haven't made the same modifications to get_triggers.

File: eldritch/events.py (hunk 14 above)
Comment (fingoldfish) [reply eldritch/events.py:2449]: This nitpick and others in events.py look like they got missed?

File: eldritch/test_events.py (hunk 17 above)
Comment (fingoldfish) [reply eldritch/test_events.py:313]: Would still like for the `moved` attribute to be tested.

**Review by fingoldfish — APPROVED** [review]
